### PR TITLE
Compute report lazily

### PR DESCRIPTION
**Pull request by fatkodima**
Currently, `memory_profiler` calculates report eagerly, which wastes time/memory and may not always be needed. 

For example, I usually disable strings report by `.pretty_print(retained_strings: 0, allocated_strings: 0)`. 
Or someone can use `.pretty_print(detailed_report: false)`.

In this PR, report is computed lazily.

Tested on `rubocop`'s codebase (ran rubocop on its own `lib/rubocop/cop/style` files - 222 files).

### Before
```
222 files inspected, no offenses detected
Building memory report...
Finished in 77.41 seconds
```

### After
```
222 files inspected, no offenses detected
Building memory report...
Finished in 54.59 seconds
```

### Before
```
Total allocated: 47.76 MB (305368 objects)
Total retained:  508.08 kB (8818 objects)

allocated memory by gem
-----------------------------------
  47.76 MB  memory_profiler/lib

allocated memory by file
-----------------------------------
  43.79 MB  /Users/fatkodima/Desktop/oss/memory_profiler/lib/memory_profiler/results.rb
   3.96 MB  /Users/fatkodima/Desktop/oss/memory_profiler/lib/memory_profiler/top_n.rb
  728.00 B  /Users/fatkodima/Desktop/oss/memory_profiler/lib/memory_profiler/reporter.rb
...
```

### After
```
Total allocated: 9.81 kB (19 objects)
Total retained:  352.00 B (3 objects)

allocated memory by gem
-----------------------------------
   9.81 kB  memory_profiler/lib

allocated memory by file
-----------------------------------
   9.22 kB  /Users/fatkodima/Desktop/oss/memory_profiler/lib/memory_profiler/results.rb
  584.00 B  /Users/fatkodima/Desktop/oss/memory_profiler/lib/memory_profiler/reporter.rb
...
```

### CODE REVIEW

**Comment by SamSaffron** [comment]
I think some of the metric were calculated together, is there any impact, performance wise adding cost when you are displaying everything or is the "display everything" performance not impacted?

**Comment by fatkodima** [comment]
The "display everything" performance should not be impacted.
Just to be sure, I rerun my testing on both original code and this PR and got the same time.

**Comment by fatkodima** [comment]
@SamSaffron Wdyt about this?

**Comment by SamSaffron** [comment]
I do support this change I worry a bit about the extra internal complexity, cause it is a bit hard to follow all the meta programming, that said, the perf gain is unquestionably good. ... Lets try it.
